### ISSUE TITLE
RyuJIT can't seem to make up its mind about sizeof(bool)

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#446]
When bools are produced their sizeof is 4 but when bools are consumed the JIT compiler seems to think that sizeof is 1 and zero expands the bools.

```C#
using System.Runtime.CompilerServices;
class BoolSize {
    static int Main() {
        return DynamicBool(0) == StaticBool() ? 42 : 314;
    }
    [MethodImpl(MethodImplOptions.NoInlining)]
    private static bool StaticBool() {
        return true;
    }
    [MethodImpl(MethodImplOptions.NoInlining)]
    private static bool DynamicBool(int i) {
        return i.Equals(0);
    }
}
```

Code generated by RyuJIT:
```
00007FFC60B90485 33 C9                xor         ecx,ecx  
00007FFC60B90487 E8 FC FB FF FF       call        00007FFC60B904D0  
00007FFC60B9048C 40 0F B6 F0          movzx       esi,al  ; zero expand bool
00007FFC60B90490 E8 EB FB FF FF       call        00007FFC60B904F0  
00007FFC60B90495 25 FF 00 00 00       and         eax,0FFh ; zero expand bool
00007FFC60B9049A 3B F0                cmp         esi,eax  
00007FFC60B9049C 74 0B                je          00007FFC60B904A9  
...

00007FFC60B904D0 85 C9                test        ecx,ecx  
00007FFC60B904D2 0F 94 C0             sete        al  
00007FFC60B904D5 0F B6 C0             movzx       eax,al  
00007FFC60B904D8 C3                   ret  

00007FFC60B904F0 B8 01 00 00 00       mov         eax,1  
00007FFC60B904F5 C3                   ret  
```

**Comment by jkotas** [comment]
@cmckinsey Could somebody on your team please take a look?

**Comment by cmckinsey** [comment]
Thanks. Agreed the JIT should be consistent about when it does extensions. For native code, Microsoft code-generators follow a producer extension model. So for this case the callee must guarantee a fully-extended return value even for small types. This means the caller can then assume this and optimize accordingly.

For managed code, I'm pretty sure we want to use the same convention. However for compatibility and bring-up reasons for RyuJIT, we are generating both producer and consumer extensions. Look in importer.cpp line 5459 with the relevant comment. We had hit some bugs in the past in this area during bring-up of the new JIT and extending on both sides was the conservative way to get correctness.

We should probably disable this for FEATURE_CORECLR since it's not relevant (we do not provide the old JIT as a back-up scenario). This would likely result in nice code size wins.

Feel free to experiment and/or someone on the team will grab this when we get a chance. 

Thanks.

**Comment by cmckinsey** [comment]
Made the change locally and confirmed the code is optimized as expected:

G_M34988_IG01:
       56                   push     rsi
       4883EC20             sub      rsp, 32

G_M34988_IG02:
       33C9                 xor      ecx, ecx
       E8FCFBFFFF           call     BoolSize:DynamicBool(int):bool
       8BF0                 mov      esi, eax
       E8EDFBFFFF           call     BoolSize:StaticBool():bool
       3BF0                 cmp      esi, eax
       740B                 je       SHORT G_M34988_IG04
       B83A010000           mov      eax, 314

G_M34988_IG03:
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret      

G_M34988_IG04:
       B82A000000           mov      eax, 42

G_M34988_IG05:
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret

**Comment by mikedn** [comment]
>We should probably disable this for FEATURE_CORECLR since it's not relevant (we do not provide the old JIT as a back-up scenario). This would likely result in nice code size wins.

Seems like a good compromise. When I created this I thought it's just an oversight in "transforming" the x86 JIT in RyuJIT (x86 doesn't have this issue), I didn't realize that JIT64 doesn't expand the bools like x86 does.

**Comment by cmckinsey** [comment]
I'm looking at the test and perf impact of this change now. More soon.

**Comment by cmckinsey** [comment]
Fixed with PR #654
